C7Engine/MapUnitExtensions.cs

```diff
@@ -102,53 +97,111 @@ public static void RollToPromote(this MapUnit unit, MapUnit opponent, bool waitF
+
+			defensiveBombarder.animate(MapUnit.AnimatedAction.ATTACK1, true);
+
+			// dADB = defense Against Defensive Bombard
+			double dADB = attacker.StrengthVersus(defensiveBombarder, CombatRole.DefensiveBombardDefense, defensiveBombarder.facingDirection);
+			if (EngineStorage.gameData.rng.NextDouble() < defensiveBombarderStrength / (defensiveBombarderStrength + dADB))
+				attacker.hitPointsRemaining -= 1;
+
+			defensiveBombarder.defensiveBombardsRemaining -= 1;
+			defensiveBombarder.facingDirection = dBOriginalDirection;
+		}
+
+		// TODO: Check these rules are accurate and that we're not missing any.
+		bool defenderEligibleToRetreat = defender.hitPointsRemaining > 1 && ! defender.location.HasCity;
 
 		// Do combat rounds
-		while ((unit.hitPointsRemaining > 0) && (defender.hitPointsRemaining > 0)) {
+		while (true) {
 			defender.animate(MapUnit.AnimatedAction.ATTACK1, false);
-			unit    .animate(MapUnit.AnimatedAction.ATTACK1, true );
-			if (EngineStorage.gameData.rng.NextDouble() < attackerOdds)
+			attacker.animate(MapUnit.AnimatedAction.ATTACK1, true );
+			if (EngineStorage.gameData.rng.NextDouble() < attackerOdds) {
+				if (defenderEligibleToRetreat &&
+				    defender.hitPointsRemaining == 1 &&
+				    EngineStorage.gameData.rng.NextDouble() < defender.RetreatChance(attacker, false)) {
+					Tile retreatDestination = defender.location.neighbors[attacker.facingDirection];
+					if ((retreatDestination != Tile.NONE) && defender.CanEnterTile(retreatDestination, false)) {
+						defender.move(attacker.facingDirection, true);
+						result = CombatResult.DefenderRetreated;
+						break;
+					}
+				}
 				defender.hitPointsRemaining -= 1;
-			else
-				unit.hitPointsRemaining -= 1;
+				if (defender.hitPointsRemaining <= 0) {
+					result = CombatResult.DefenderKilled;
+					break;
+				}
+			} else {
+				if (attacker.hitPointsRemaining == 1 &&
+				    EngineStorage.gameData.rng.NextDouble() < attacker.RetreatChance(defender, true)) {
+					result = CombatResult.AttackerRetreated;
+					break;
+				}
+				attacker.hitPointsRemaining -= 1;
+				if (attacker.hitPointsRemaining <= 0) {
+					result = CombatResult.AttackerKilled;
+					break;
+				}
+			}
 		}
 
-		MapUnit loser = (defender.hitPointsRemaining <= 0) ? defender : unit,
-			winner = (defender == loser) ? unit : defender;
-
-		winner.RollToPromote(loser, false);
-
-		// Play death animation
-		loser.animate(MapUnit.AnimatedAction.DEATH, true);
-		loser.disband();
+		if ((result == CombatResult.AttackerKilled) || (result == CombatResult.DefenderKilled)) {
+			var (dead, alive) = (result == CombatResult.AttackerKilled) ? (attacker, defender) : (defender, attacker);
+			alive.RollToPromote(dead, false);
+			dead.animate(MapUnit.AnimatedAction.DEATH, true);
+			dead.disband();
+		}
 
-		if (defender != loser)
+		if (result.DefenderWon())
 			defender.facingDirection = defenderOriginalDirection;
 
-		return unit != loser;
+		return result;
```

Comment:
Probably for the next set of improvements, but at over 100 lines of code and now covering several (albeit related) mechanics, I'm thinking it's about time to add helper methods.  E.g. defensive bombard could probably be broken out.  I found as I added more code to my editor that it's easier to find the relevant section for making updates if methods are split out at least somewhat, especially when using an IDE that gives an outline of the available methods.